geopyspark-backend/geotrellis/src/main/scala/geopyspark/geotrellis/SpatialTiledRasterLayer.scala

```diff
@@ -210,6 +210,31 @@ class SpatialTiledRasterLayer(
   def mask(geometries: Seq[MultiPolygon]): TiledRasterLayer[SpatialKey] =
     SpatialTiledRasterLayer(zoomLevel, Mask(rdd, geometries, Mask.Options.DEFAULT))
 
+  def mask(
+    aggregatedRDD: RDD[(SpatialKey, Seq[Geometry])],
+    producePartitioner: () => Option[Partitioner],
+    options: Rasterizer.Options
+  ): TiledRasterLayer[SpatialKey] = {
+    val mapTrans = rdd.metadata.layout.mapTransform
+
+    val joinedRDD: RDD[(SpatialKey, (MultibandTile, Seq[Geometry]))] =
+      producePartitioner() match {
+        case Some(p) =>
+          rdd.join(aggregatedRDD, p)
+        case None =>
+          rdd.join(aggregatedRDD)
+      }
+
+    val maskedRDD: RDD[(SpatialKey, MultibandTile)] =
+      joinedRDD.mapPartitions ({ partition =>
```

Comment:
We really should just make `.mapValuesWithKey` that implements this `.mapPartitions` `preservePartitioning` pattern as an implicit extension like `PairRDDFunctions`, we just do it so often.

Reply:
That's a good point. I noticed that we do use this convention a lot throughout our code bases.